src/main/java/sx/blah/discord/util/MessageHistoryBuilder.java

```diff
@@ -0,0 +1,429 @@
+	 *
+	 * @param endID the message ID to end at
+	 * @return this, for chaining
+	 */
+	public MessageHistoryBuilder endAt(long endID) {
+		this.uEnd = channel.getMessageByID(endID);
+		if (uEnd == null)
+			throw new IllegalArgumentException("Message " + Long.toUnsignedString(endID) + " does not exist in channel " + channel.getStringID() + "!");
+		this.uEndTime = uEnd.getTimestamp();
+		this.uIncludeEnd = false;
+		return this;
+	}
+
+	/**
+	 * Specifies the message to end at (exclusive).
+	 *
+	 * @param msg the message ID to end at
+	 * @return this, for chaining
+	 */
+	public MessageHistoryBuilder endAt(IMessage msg) {
+		if (!channel.equals(msg.getChannel()))
+			throw new IllegalArgumentException("Message " + msg.getStringID() + " does not exist in channel " + channel.getStringID() + "!");
+		this.uEnd = msg;
+		this.uEndTime = msg.getTimestamp();
+		this.uIncludeEnd = false;
+		return this;
+	}
+
+	/**
+	 * Specifies the time to end at (exclusive).
+	 *
+	 * @param endTime the time to end at
+	 * @return this, for chaining
+	 */
+	public MessageHistoryBuilder endAt(LocalDateTime endTime) {
+		this.uEnd = null;
+		this.uEndTime = endTime;
+		this.uIncludeEnd = false;
+		return this;
+	}
+
+	/**
+	 * Specifies the message to end at.
+	 *
+	 * @param endID     the message ID to end at
+	 * @param inclusive whether or not to include this time
+	 * @return this, for chaining
+	 */
+	public MessageHistoryBuilder endAt(long endID, boolean inclusive) {
+		this.uEnd = channel.getMessageByID(endID);
+		if (uEnd == null)
+			throw new IllegalArgumentException("Message " + Long.toUnsignedString(endID) + " does not exist in channel " + channel.getStringID() + "!");
+		this.uEndTime = uEnd.getTimestamp();
+		this.uIncludeEnd = inclusive;
+		return this;
+	}
+
+	/**
+	 * Specifies the message to end at.
+	 *
+	 * @param msg       the message ID to end at
+	 * @param inclusive whether or not to include this time
+	 * @return this, for chaining
+	 */
+	public MessageHistoryBuilder endAt(IMessage msg, boolean inclusive) {
+		if (!channel.equals(msg.getChannel()))
+			throw new IllegalArgumentException("Message " + msg.getStringID() + " does not exist in channel " + channel.getStringID() + "!");
+		this.uEnd = msg;
+		this.uEndTime = msg.getTimestamp();
+		this.uIncludeEnd = inclusive;
+		return this;
+	}
+
+	/**
+	 * Specifies the time to end at.
+	 *
+	 * @param endTime   the time to end at
+	 * @param inclusive whether or not to include this time
+	 * @return this, for chaining
+	 */
+	public MessageHistoryBuilder endAt(LocalDateTime endTime, boolean inclusive) {
+		this.uEnd = null;
+		this.uEndTime = endTime;
+		this.uIncludeEnd = inclusive;
+		return this;
+	}
+
+	/**
+	 * Specifies the maximum number of messages to get. Defaults to -1 (no limit).
+	 *
+	 * @param count max number of messages to get
+	 * @return this, for chaining
+	 */
+	public MessageHistoryBuilder withMaxCount(int count) {
+		if (count < 1) throw new IllegalArgumentException();
```

Comment:
I guess not. I wasn't sure whether to include the check or not because it might help with finding logic errors (why would you call withMaxCount if you wanted a limitless range?) but could lead to unnecessarily long logic:
`.withMaxCount(noLimit? -1: 50)`
Vs
```java
<get builder>
if(limited) builder.withMaxCount(50)```
Though looking at it now, I actually think I prefer the latter because it is more readable and easier to understand